File: scratch/build_manual_figures.js
// Screenshots for manual sections 1.4, 5.4, 8.4 and 9.4, from the files in examples/.
// Writes raw captures plus element boxes to scratch/_fig/; scratch/build_manual_figures.py
// then labels and assembles them into img/.
//   node scratch/build_manual_figures.js && python scratch/build_manual_figures.py
const fs = require('fs');
const path = require('path');
const { launch } = require('../tests/lib/browser');
const { start } = require('../tests/lib/static-server');

const ROOT = path.join(__dirname, '..');
const OUT = path.join(__dirname, '_fig');
fs.mkdirSync(OUT, { recursive: true });

(async () => {
  const { server, baseUrl } = await start();
  const b = await launch();
  const newPage = async (file, vp) => { const p = await b.newPage({ viewport: vp || { width: 1400, height: 820 }, deviceScaleFactor: 2 }); const errs = []; p.on('pageerror', e => errs.push(e.message)); p._errs = errs; await p.goto(baseUrl + '/index.html', { waitUntil: 'networkidle' }); await p.setInputFiles('#fileInput', path.join(ROOT, 'examples', file)); await p.waitForTimeout(2500); return p; };
  const report = {};

  // ---- 1.4 Interface layout: svk_k4.fa, Display menu open ------------------------------
  {
    const p = await newPage('svk_k4.fa');
    await p.hover('.section-header[data-section="display"]');
    await p.waitForTimeout(500);
    const boxes = await p.evaluate(() => {
      const box = sel => { const e = document.querySelector(sel); if (!e) return null; const r = e.getBoundingClientRect(); return { x: r.left, y: r.top, w: r.width, h: r.height }; };
      // version + filename sit together in the top-right corner
      const v = document.getElementById('versionIndicator').getBoundingClientRect();
      const f = document.getElementById('sourceInfo').getBoundingClientRect();
      const x0 = Math.min(v.left, f.left), y0 = Math.min(v.top, f.top), x1 = Math.max(v.right, f.right), y1 = Math.max(v.bottom, f.bottom);
      return {
        // the whole menu row: every visible top-bar control from Input to Actions
        menus: (() => {
          const skip = new Set(['versionIndicator', 'sourceInfo']);
          const rs = [...document.querySelectorAll('#controls button, #controls .section-header')]
            .filter(e => !skip.has(e.id) && !e.closest('#versionIndicator, #sourceInfo'))
            .map(e => e.getBoundingClientRect()).filter(r => r.width > 0 && r.top < 30 && r.bottom < 40);
          const x0 = Math.min(...rs.map(r => r.left)), y0 = Math.min(...rs.map(r => r.top));
          const x1 = Math.max(...rs.map(r => r.right)), y1 = Math.max(...rs.map(r => r.bottom));
          return { x: x0 - 3, y: y0 - 3, w: x1 - x0 + 6, h: y1 - y0 + 6 };
        })(),
        openMenu: box('#display-controls'),
        modeSwitch: box('#quickModeSwitch'),
        viewport: box('#alignmentContainer'),
        versionFile: { x: x0, y: y0, w: x1 - x0, h: y1 - y0 },
        versionText: document.getElementById('versionIndicator').textContent.trim(),
        fileText: document.getElementById('sourceInfo').textContent.trim()
      };
    });
    await p.screenshot({ path: path.join(OUT, 'layout.png') });
    report.layout = boxes; report.layoutErrors = p._errs;
    await p.close();
  }

  // ---- 5.4 Codon analysis: synth_msa.fa ----------------------------------------------
  {
    const p = await newPage('synthetic/synth_msa.fa', { width: 1300, height: 600 });
    await p.evaluate(() => {
      document.getElementById('modeSingle').checked = true; onModeChange();
      // conservation shading off so the codon underlines are visible
      for (const id of ['enableBlack', 'enableDark', 'enableLight']) { const e = document.getElementById(id); e.checked = false; e.dispatchEvent(new Event('change', { bubbles: true })); }
      const c = document.getElementById('codonAnalysis'); c.checked = true; c.dispatchEvent(new Event('change', { bubbles: true }));
    });
    await p.waitForTimeout(1500);
    const info = await p.evaluate(() => ({
      codonOn: !!state._codonData, frame: state._codonActiveFrame,
      stops: state._codonData.stops.map((s, i) => [state.seqs[i].header, s.length]),
      frameshifts: state._codonData.frameShifts.map((s, i) => [state.seqs[i].header, s.length]),
      rows: state.seqs.map(s => s.header),
      // which 1-based columns the page actually marks, per row
      marked: Object.fromEntries(state.seqs.map((s, i) => [s.header, {
        syn: [...document.querySelectorAll(`.seq-line[data-seq-index="${i}"] .seq-data > span.codon-syn`)].map(e => +e.dataset.pos + 1),
        nonsyn: [...document.querySelectorAll(`.seq-line[data-seq-index="${i}"] .seq-data > span.codon-nonsyn`)].map(e => +e.dataset.pos + 1)
      }]))
    }));
    // Columns 1-60 only: the synthetic CDS is columns 1-60 (examples/synthetic/README.md);
    // past it, frame 0 reads non-coding sequence and marks stops that mean nothing
    const cr = await p.evaluate(() => {
      const c = document.getElementById('alignmentContainer').getBoundingClientRect();
      const lines = [...document.querySelectorAll('#alignmentContainer .seq-line')];
      const last = lines[lines.length - 1].getBoundingClientRect();
      const sp = document.querySelector('.seq-line[data-seq-index="0"] .seq-data > span[data-pos="59"]').getBoundingClientRect();
      return { x: c.left, y: c.top, w: sp.right - c.left + 4, h: last.bottom - c.top + 30 };
    });
    await p.screenshot({ path: path.join(OUT, 'codon.png'), clip: { x: cr.x, y: cr.y, width: cr.w, height: cr.h } });
    report.codon = info; report.codonErrors = p._errs;
    await p.close();
  }

  // ---- 8.4 Move NoGaps and Slide KeepGaps: real drags on synth_msa.fa ------------------
  {
    const p = await newPage('synthetic/synth_msa.fa', { width: 1300, height: 600 });
    await p.evaluate(() => { document.getElementById('modeSingle').checked = true; onModeChange(); });
    await p.waitForTimeout(800);
    await p.click('#editToggleButton');
    await p.waitForTimeout(300);
    const clip = async (name) => {
      // first 30 columns of all six rows, names included
      const r = await p.evaluate(() => {
        const rows = [...document.querySelectorAll('.seq-line[data-seq-index]')];
        const first = rows[0].getBoundingClientRect(), last = rows[rows.length - 1].getBoundingClientRect();
        const span30 = rows[0].querySelector('.seq-data > span[data-pos="29"]').getBoundingClientRect();
        return { x: Math.max(0, first.left - 2), y: first.top - 3, w: span30.right - first.left + 6, h: last.bottom - first.top + 6 };
      });
      await p.screenshot({ path: path.join(OUT, name), clip: { x: r.x, y: r.y, width: r.w, height: r.h } });
    };
    const drag = async (tool, row, pos, cols) => {
      await p.click(tool === 'moveNoGaps' ? '#editMoveNoGapsButton' : '#editSlideKeepGapsButton');
      await p.waitForTimeout(200);
      const s = await p.evaluate(({ row, pos }) => {
        const sp = document.querySelector(`.seq-line[data-seq-index="${row}"] .seq-data > span[data-pos="${pos}"]`);
        const r = sp.getBoundingClientRect(); return { x: r.left + r.width / 2, y: r.top + r.height / 2, w: r.width };
      }, { row, pos });
      await p.mouse.move(s.x, s.y); await p.mouse.down();
      const steps = Math.abs(cols) * 4;
      for (let i = 1; i <= steps; i++) { await p.mouse.move(s.x + (cols * s.w) * i / steps, s.y); await p.waitForTimeout(15); }
      await p.mouse.up();
      await p.mouse.move(5, 590); await p.waitForTimeout(700); // off the alignment, so no tooltip
    };
    const seqOf = (i) => p.evaluate(i => state.seqs[i].seq.slice(0, 30), i);
    const rows = await p.evaluate(() => state.seqs.map(s => s.header));
    const del = rows.indexOf('beta_del'), fs1 = rows.indexOf('beta_fs');
    report.move = { row: 'beta_del', before: await seqOf(del) };
    await clip('move_before.png');
    // grab the residue at column 13 (0-based 12) and drag 3 columns right: closes the 3-base gap
    await drag('moveNoGaps', del, 12, 3.5);
    report.move.after = await seqOf(del);
    await clip('move_after.png');
    report.slide = { row: 'beta_fs', before: await seqOf(fs1) };
    // Slide KeepGaps: grab column 6 (0-based 5), just right of the one-base gap, drag 1 left
    await drag('slideKeepGaps', fs1, 5, -1);
    report.slide.after = await seqOf(fs1);
    await clip('slide_after.png');
    report.editErrors = p._errs;
    await p.close();
  }

  // ---- 9.4 Tree window: 12 SVK copies, three per subfamily ---------------------------
  {
    const p = await newPage('svk_k4.fa', { width: 1300, height: 900 });
    const picked = await p.evaluate(() => {
      const want = []; const byGroup = {};
      state.seqs.forEach((s, i) => { const g = s.header.slice(0, 2); (byGroup[g] ||= []).push(i); });
      Object.values(byGroup).forEach(ix => want.push(...ix.slice(0, 3)));
      state.selectedRows = new Set(want);
      openTreeBuilder();
      return want.map(i => state.seqs[i].header);
    });
    await p.waitForTimeout(1500);
    const tree = await p.evaluate(() => ({ summary: document.getElementById('treeBuilderSummary').textContent, newick: document.getElementById('treeNewickOutput').value.slice(0, 200), svg: !!document.querySelector('#treeBuilderDialog svg') }));
    // title, toolbar and drawing only: the dialog is taller than the window
    const tc = await p.evaluate(() => {
      const d = document.getElementById('treeBuilderDialog').getBoundingClientRect();
      const svg = document.querySelector('#treeBuilderDialog svg');
      let box = svg.parentElement; // the resizable drawing box around the svg
      const r = box.getBoundingClientRect();
      return { x: d.left, y: d.top, w: d.width, h: r.bottom - d.top + 8 };
    });
    await p.screenshot({ path: path.join(OUT, 'tree.png'), clip: { x: tc.x, y: tc.y, width: tc.w, height: tc.h } });
    report.tree = { picked, ...tree }; report.treeErrors = p._errs;
    await p.close();
  }

  fs.writeFileSync(path.join(OUT, 'report.json'), JSON.stringify(report, null, 1));
  console.log(JSON.stringify(report, null, 1));
  await b.close(); server.close();
})();

File: scratch/build_manual_figures.py
"""Assemble manual figures from the captures made by build_manual_figures.js.

Reads scratch/_fig/*.png and report.json; writes img/interface-layout.png,
img/codon-example.png, img/move-slide-example.png and img/tree-example.png.
Run: node scratch/build_manual_figures.js && python scratch/build_manual_figures.py
"""
import json
import os
import shutil

from PIL import Image, ImageDraw, ImageFont

HERE = os.path.dirname(__file__)
FIG = os.path.join(HERE, '_fig')
IMG = os.path.join(HERE, '..', 'img')
SCALE = 2  # deviceScaleFactor used for the captures
report = json.load(open(os.path.join(FIG, 'report.json'), encoding='utf-8'))


def font(size):
    for name in ('arialbd.ttf', 'DejaVuSans-Bold.ttf'):
        try:
            return ImageFont.truetype(name, size)
        except OSError:
            pass
    return ImageFont.load_default()


# ---- 1.4 layout: numbered outlines around the four areas ------------------------------
im = Image.open(os.path.join(FIG, 'layout.png')).convert('RGB')
d = ImageDraw.Draw(im)
L = report['layout']
areas = [('1', L['menus']), ('2', L['openMenu']), ('3', L['viewport']), ('4', L['versionFile'])]
colour = (220, 40, 40)
f = font(34)
for label, b in areas:
    x0, y0 = b['x'] * SCALE, b['y'] * SCALE
    x1, y1 = (b['x'] + b['w']) * SCALE, min((b['y'] + b['h']) * SCALE, im.height - 3)
    d.rectangle([x0, y0, x1, y1], outline=colour, width=5)
    # badge inside the top-left corner of large areas; just below thin ones (1 and 4)
    # so it covers no control
    bx, by = (x0 + 8, y1 + 6) if label in ('1', '4') else (x0 + 8, y0 + 8)
    if label == '1':   # empty stretch of the row, between Alignment and Keys
        bx = x0 + (x1 - x0) * 0.78
    if label == '3':   # blank bottom-right corner of the viewport
        bx, by = x1 - 60, y1 - 60
    d.ellipse([bx, by, bx + 44, by + 44], fill=colour)
    d.text((bx + 22, by + 22), label, fill='white', font=f, anchor='mm')
im.save(os.path.join(IMG, 'interface-layout.png'), optimize=True)

# ---- 5.4 codon and 9.4 tree: used as captured -----------------------------------------
shutil.copyfile(os.path.join(FIG, 'codon.png'), os.path.join(IMG, 'codon-example.png'))
shutil.copyfile(os.path.join(FIG, 'tree.png'), os.path.join(IMG, 'tree-example.png'))

# ---- 8.4 move/slide: three labelled panels --------------------------------------------
panels = [('Before', 'move_before.png'),
          ('After Move NoGaps on beta_del (grab column 13, drag 3 right)', 'move_after.png'),
          ('After Slide KeepGaps on beta_fs (grab column 6, drag 1 left)', 'slide_after.png')]
ims = [Image.open(os.path.join(FIG, n)).convert('RGB') for _, n in panels]
lab = font(24)
d0 = ImageDraw.Draw(Image.new('RGB', (1, 1)))
pad, head = 16, 40
W = max(max(i.width for i in ims), max(int(d0.textlength(tt, font=lab)) for tt, _ in panels)) + 2 * pad
H = sum(i.height + head + pad for i in ims) + pad
out = Image.new('RGB', (W, H), 'white')
d = ImageDraw.Draw(out)
y = pad
for (title, _), i in zip(panels, ims):
    d.text((pad, y + 6), title, fill=(40, 40, 40), font=lab)
    out.paste(i, (pad, y + head))
    y += head + i.height + pad
out.save(os.path.join(IMG, 'move-slide-example.png'), optimize=True)

for n in ('interface-layout', 'codon-example', 'move-slide-example', 'tree-example'):
    print(n, Image.open(os.path.join(IMG, n + '.png')).size)
